apisix/plugins/ai-proxy.lua

```diff
@@ -0,0 +1,136 @@
+local ngx_flush = ngx.flush
+
+local plugin_name = "ai-proxy"
+local _M = {
+    version = 0.5,
+    priority = 1004,
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ai_driver = pcall(require, "apisix.plugins.ai-proxy.drivers." .. conf.model.provider)
+    if not ai_driver then
+        return false, "provider: " .. conf.model.provider .. " is not supported."
+    end
+    return core.schema.check(schema.plugin_schema, conf)
+end
+
+
+local CONTENT_TYPE_JSON = "application/json"
+
+
+local function keepalive_or_close(conf, httpc)
+    if conf.set_keepalive then
+        httpc:set_keepalive(10000, 100)
+        return
+    end
+    httpc:close()
+end
+
+
+function _M.access(conf, ctx)
+    local ct = core.request.header(ctx, "Content-Type") or CONTENT_TYPE_JSON
+    if not core.string.has_prefix(ct, CONTENT_TYPE_JSON) then
+        return 400, "unsupported content-type: " .. ct
+    end
+
+    local request_table, err = core.request.get_request_body_table()
+    if not request_table then
+        return 400, err
+    end
+
+    local ok, err = core.schema.check(schema.chat_request_schema, request_table)
+    if not ok then
+        return 400, "request format doesn't match schema: " .. err
+    end
+
+    if conf.model.name then
+        request_table.model = conf.model.name
+    end
+
+    if core.table.try_read_attr(conf, "model", "options", "stream") then
+        request_table.stream = true
+    end
+
+    local ai_driver = require("apisix.plugins.ai-proxy.drivers." .. conf.model.provider)
+    local res, err, httpc = ai_driver.request(conf, request_table, ctx)
+    if not res then
+        core.log.error("failed to send request to AI service: ", err)
+        return 500
+    end
+
+    local body_reader = res.body_reader
+    if not body_reader then
+        core.log.error("LLM sent no response body")
+        return 500
+    end
+
+    if conf.passthrough then
+        ngx_req.init_body()
+        while true do
+            local chunk, err = body_reader() -- will read chunk by chunk
+            if err then
+                core.log.error("failed to read response chunk: ", err)
+                break
+            end
+            if not chunk then
+                break
+            end
+            ngx_req.append_body(chunk)
+        end
+        ngx_req.finish_body()
+        keepalive_or_close(conf, httpc)
+        return
+    end
+
+    if core.table.try_read_attr(conf, "model", "options", "stream") then
+        while true do
+            local chunk, err = body_reader() -- will read chunk by chunk
+            if err then
+                core.log.error("failed to read response chunk: ", err)
+                break
+            end
+            if not chunk then
+                break
+            end
+            ngx_print(chunk)
+            ngx_flush(true)
+        end
+        keepalive_or_close(conf, httpc)
+        return
+    else
+        local res_body, err = res:read_body()
```

Comment:
also using body_reader, ngx_print * N + ngx_flush * 1?

Reply:
I don't understand what you mean.

Reply:
The purpose is to save the intermediate buffering process and directly
write the response body to the response buffer by calling ngx.print
multiple times. Finally, call ngx.flush once more.

But,

$ restydoc -s ngx.print

> Please note that both "ngx.print" and ngx.say will always invoke the
> whole Nginx output body filter chain, which is an expensive operation.
> So be careful when calling either of these two in a tight loop; buffer
> the data yourself in Lua and save the calls.

Reply:
got it, I think we don't need ` ngx_print * N + ngx_flush * 1` as reading the response body all at once is favourable for `"stream": false`

Reply:
Hi @shreemaan-abhishek Is it resolved?

Reply:
yes